### PR TITLE
dev/core#2159 Handle exceptions in Mail:send class

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
In theory the mail class can be easily swapped out - but expecting a replacement to return a PEAR_ERROR is a bit much! This adds handling for exceptions

https://lab.civicrm.org/dev/core/-/issues/2159

Before
----------------------------------------
PEAR_Error is required on error

After
----------------------------------------
An exception can be thrown

Technical Details
----------------------------------------
Mostly in https://lab.civicrm.org/dev/core/-/issues/2159

Comments
----------------------------------------
- need to think a bit more about the messages though

### CODE REVIEW

**Comment by mattwire** [comment]
This is kind of related to https://github.com/civicrm/civicrm-core/pull/18466 and should probably be reviewed in parallel

**Comment by monishdeb** [comment]
I am happy with the fix. @eileenmcnaughton can you add a unit test which try-catch the expection and assert the error message for using wrong paramter or something else which could result in an exception?

**Comment by eileenmcnaughton** [comment]
@monishdeb that was a bit of a curly one to write a test for - I wasn't sure for a minute there I could figure it out but I'm happy with the result - as long as jenkins is!

**Comment by mattwire** [comment]
Linking to https://github.com/civicrm/civicrm-core/pull/18466

**Comment by eileenmcnaughton** [comment]
Test fail was because it couldn't serialize my setting (because it had an object in it). I think it's fixed. In terms of how this helps with #18466 - it removes one of the steps en-route to removing ignoreException mode. If we could be sure that we were always throwing exceptions (the thing ignoreException mode messes with) we could be sure that #18466 is safe

File: CRM/Utils/Mail.php [hunk 1]
```diff
@@ -301,8 +302,13 @@ public static function send(&$params) {
     }
 
     if (is_object($mailer)) {
-      $errorScope = CRM_Core_TemporaryErrorScope::ignoreException();
-      $result = $mailer->send($to, $headers, $message);
+      try {
+        $result = $mailer->send($to, $headers, $message);
+      }
+      catch (Exception $e) {
+        CRM_Core_Session::setStatus($e->getMessage(), ts('Mailing Error'), 'error');
```
Comment (seamuslee001) [inline CRM/Utils/Mail.php:309]: is there a reason why your not logging this via errorMessage?

Comment (eileenmcnaughton) [reply CRM/Utils/Mail.php:309]: It's consisent with a couple of lines lower down / what happened before

Comment (seamuslee001) [reply CRM/Utils/Mail.php:309]: Doh thought it was a logger np then

**Comment by eileenmcnaughton** [comment]
@seamuslee001 yay it passed

**Comment by seamuslee001** [comment]
This makes sense to me and ensures we are generating a consistent return and I did some r-run and confirmed the backtrace didn't seem to change and this functionally ensures that we are returning the same thing if our source threw an Exception or a Pear Error, I'm going to merge this as I think this is a good improvement

**Pull request merged**